File: LeetCode-C++/LeetCode-C++/二叉树/_173_二叉搜索树迭代器.cpp
//
//  _173_二叉搜索树迭代器.cpp
//  LeetCode-C++
//
//  Created by warden on 2020/2/17.
//  Copyright © 2020 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/binary-search-tree-iterator/

#include <stdio.h>
#include <stack>
#include <vector>
#include "TreeNode.hpp"

using namespace std;

class BSTIterator {
private:
    stack<TreeNode*> st;
public:
    BSTIterator(TreeNode* root) {
        while (root) {
            st.push(root);
            root = root->left;
        }
    }
    
    /** @return the next smallest number */
    int next() {
        TreeNode *node = st.top();
        st.pop();
        int res = node->val;
        node = node->right;
        while (node) {
            st.push(node);
            node = node->left;
        }
        return res;
    }
    
    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !st.empty();
    }
};

/**
 示例：
        7
       /\
      3  15
         /\
        9 20


 BSTIterator iterator = new BSTIterator(root);
 iterator.next();    // 返回 3
 iterator.next();    // 返回 7
 iterator.hasNext(); // 返回 true
 iterator.next();    // 返回 9
 iterator.hasNext(); // 返回 true
 iterator.next();    // 返回 15
 iterator.hasNext(); // 返回 true
 iterator.next();    // 返回 20
 iterator.hasNext(); // 返回 false
 */
/**
int main(int argc, const char * argv[]) {
    vector<int> nodes = {7,3,15,INT_MAX,INT_MAX,9,20};
    TreeNode *root = generateTreeByLevel(nodes);
    BSTIterator* iterator = new BSTIterator(root);
    printf("%d\n", iterator->next());   // 返回3
    printf("%d\n", iterator->next());   // 返回7
    printf("%d\n", iterator->hasNext());// 返回true
    printf("%d\n", iterator->next());   // 返回9
    printf("%d\n", iterator->hasNext());// 返回true
    printf("%d\n", iterator->next());   // 返回15
    printf("%d\n", iterator->hasNext());// 返回true
    printf("%d\n", iterator->next());   // 返回20
    printf("%d\n", iterator->hasNext());// 返回false
    return 0;
}
 */

File: LeetCode-Swift/LeetCode-Swift/二叉树/_173_二叉搜索树迭代器.swift
//
//  _173_二叉搜索树迭代器.swift
//  LeetCode-Swift
//
//  Created by warden on 2020/2/17.
//  Copyright © 2020 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/binary-search-tree-iterator/

import Foundation

class _173_二叉搜索树迭代器 {
    private var st = [TreeNode]()
    
    init(_ root: TreeNode?) {
        var node = root
        while node != nil {
            st.append(node!)
            node = node?.left
        }
    }
    
    /** @return the next smallest number */
    func next() -> Int {
        var node: TreeNode?
        node = st.removeLast()
        let res = node?.val ?? 0;
        node = node?.right
        while node != nil {
            st.append(node!);
            node = node?.left
        }
        return res;
    }
    
    /** @return whether we have a next smallest number */
    func hasNext() -> Bool {
        return !st.isEmpty
    }
}

/**
 示例：
        7
       /\
      3  15
         /\
        9 20


 BSTIterator iterator = new BSTIterator(root);
 iterator.next();    // 返回 3
 iterator.next();    // 返回 7
 iterator.hasNext(); // 返回 true
 iterator.next();    // 返回 9
 iterator.hasNext(); // 返回 true
 iterator.next();    // 返回 15
 iterator.hasNext(); // 返回 true
 iterator.next();    // 返回 20
 iterator.hasNext(); // 返回 false
 */
func test_173() {
    let root = generateTreeByLevel([7,3,15,nil,nil,9,20]);
    
    let solution = _173_二叉搜索树迭代器(root);
    print(solution.next())      // 返回3
    print(solution.next())      // 返回7
    print(solution.hasNext())   // 返回true
    print(solution.next())      // 返回9
    print(solution.hasNext())   // 返回true
    print(solution.next())      // 返回15
    print(solution.hasNext())   // 返回true
    print(solution.next())      // 返回20
    print(solution.hasNext())   // 返回false
}
